daffodil-runtime1/src/main/java/org/apache/daffodil/udf/UDFunctionService.java

```diff
@@ -0,0 +1,137 @@
+	private ArrayList<String> warnings = new ArrayList<>();
+
+	public UDFunctionService(ClassLoader cl) {
+		try {
+			loader = cl == null ? ServiceLoader.load(UDFunctionProvider.class) : ServiceLoader.load(UDFunctionProvider.class, cl);
+
+			loader.forEach(fcp -> {
+				Class<?>[] fcpfc = fcp.getFunctionClasses();
+
+				if (fcpfc == null || fcpfc.length == 0) {
+					warnings.add(String.format("Provider ignored: [%s].\nNo Function Classes found.",
+							fcp.getClass().getName()));
+					return;
+				}
+
+				List<Class<?>> fc = Arrays.asList(fcpfc);
+
+				List<Class<?>> missingAnnotations = fc.stream()
+						.filter(f -> !f.isAnnotationPresent(FunctionClassInfo.class)).collect(Collectors.toList());
+
+				List<Class<?>> nonSerializables = fc.stream().filter(f -> Serializable.class.isAssignableFrom(f))
+						.collect(Collectors.toList());
+
+				if (!missingAnnotations.isEmpty()) {
+					warnings.add(
+							String.format("Provider ignored: [%s].\nAnnotations missing for FunctionClass(es):\n%s",
+									fcp.getClass().getName(), missingAnnotations.stream().map(c -> c.getName())
+											.collect(Collectors.joining("\n", "\t[", "]"))));
+					return;
+				}
+
+				if (!nonSerializables.isEmpty()) {
+					warnings.add(String.format(
+							"Provider ignored: [%s].\nFunctionClass(es) must implement java.io.Serializable:\n%s",
+							fcp.getClass().getName(), nonSerializables.stream().map(c -> c.getName())
+									.collect(Collectors.joining("\n", "\t[", "]"))));
+					return;
+				}
+
+				fc.stream().forEach(f -> {
+					FunctionClassInfo fInfo = f.getAnnotation(FunctionClassInfo.class);
+					String ns = fInfo.namespace();
+					String fname = fInfo.name();
+					if (ns.isEmpty() || fname.isEmpty() || ns.matches("\\s+") || fname.matches("\\s+")) {
+						warnings.add(String.format("FunctionClass ignored: [%s]. From Provider [%s].%s%s", f.getName(), fcp.getClass().getName(),
+								ns.isEmpty() ? "\nAnnotation namespace field is empty or invalid." : "",
+								fname.isEmpty() ? "\nAnnotation name field is empty or invalid." : ""));
+					}
+
+					String key = String.join("_", fInfo.namespace(), fInfo.name());
+					if (functionProviderLookup.containsKey(key)) {
+						warnings.add(String.format(
+								"FunctionClass ignored: [%s].\nDuplicate name [%s] and namespace[%s] found.",
+								f.getName(), key.split("_")[0], key.split("_")[1]));
+					} else {
+						functionProviderLookup.put(key, fcp);
+						functionClasses.add(f);
+					}
+				});
+			});
+
+			if (functionClasses.isEmpty()) {
+				errors.add(String.format("No user defined functions found."
+						+ "\nCheck that UDF JARs are on classpath and that they are properly registerable by ServiceLoader."
+						+ "\nCurrent classpath locations:\n%s",
+						Arrays.asList(Misc.getClassPath()).stream().map(u -> u.toString())
+								.collect(Collectors.joining("\n"))));
+			}
+		} catch (ServiceConfigurationError | Exception e) {
+			e.printStackTrace();
```

Comment:
remove printStackTrace. You shouldn't catch Exception, well really, ever pretty much. What are the other kinds of exceptions you were hoping to catch. You should be more specific.

Reply:
I think I was also concerned about a NullPointerException here. I'll replace the general Exception with that.

Reply:
We should be checking null appropriately if there's a null pointer exception. I would avoid catchting that exception.

Reply:
Ok, sounds good!